spark/src/test/scala/org/apache/spark/sql/delta/DeltaInsertIntoImplicitCastSuite.scala

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.delta
+
+import org.apache.spark.sql.delta.sources.DeltaSQLConf
+
+import org.apache.spark.sql.SaveMode
+import org.apache.spark.sql.types._
+
+/**
+ * Test suite covering implicit casting in INSERT operations when the type of the data to insert
+ * doesn't match the type in Delta table.
+ *
+ * The casting behavior is (unfortunately) dependent on the API used to run the INSERT, e.g.
+ * Dataframe V1 insertInto() vs V2 saveAsTable() or using SQL.
+ * This suite intends to exhaustively cover all the ways INSERT can be run on a Delta table. See
+ * [[DeltaInsertIntoTest]] for a list of these INSERT operations covered.
+ */
+class DeltaInsertIntoImplicitCastSuite extends DeltaInsertIntoTest {
+
+  for (schemaEvolution <- BOOLEAN_DOMAIN) {
+    testInserts("insert with implicit up and down cast on top-level fields, " +
```

Comment:
nit: isn't there _no_ implicit up and down cast since the schema remains the same?

Reply:
The schema of the data inserted isn't the same as the schema of the table:
table: `a long, b int`
data: `a int, b long`

so there' is an upcast for `a` and downcast for `b`